vmdb/app/controllers/api_controller/vms.rb

```diff
@@ -117,5 +128,23 @@ def set_owner_vm(vm, owner)
     rescue => err
       action_result(false, err.to_s)
     end
+
+    def add_lifecycle_event_vm(vm, lifecycle_event)
+      desc = "#{vm_ident(vm)} adding lifecycle event=#{lifecycle_event[:event]} message=#{lifecycle_event[:message]}"
+      event = LifecycleEvent.create_event(vm, lifecycle_event)
+      action_result(event.present?, desc)
+    rescue => err
+      action_result(false, err.to_s)
+    end
+
+    def lifecycle_event_from_data(data)
+      data = {} if data.nil?
+      {
+        :event      => data["event"].to_s,
+        :status     => data["status"].to_s,
+        :message    => data["message"].to_s,
+        :created_by => data["created_by"].to_s
+      }
+    end
```

Comment:
This can be done simpler, I think...and the symbolized keys are not required by AR.

```ruby
data = data.slice("event", "status", "message", "created_by")
data.keys.each { |k| data[k] = data[k].to_s } 
```